client/replication.go

```diff
@@ -0,0 +1,130 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"strings"
+
+	"github.com/opensds/opensds/pkg/model"
+	"github.com/opensds/opensds/pkg/utils/urls"
+)
+
+type ReplicationBuilder *model.ReplicationSpec
+type FailoverReplicationBuilder *model.FailoverReplicationSpec
+
+// NewReplicationMgr
+func NewReplicationMgr(r Receiver, edp string, tenantId string) *ReplicationMgr {
+	return &ReplicationMgr{
+		Receiver: r,
+		Endpoint: edp,
+		TenantId: tenantId,
+	}
+}
+
+// ReplicationMgr
+type ReplicationMgr struct {
+	Receiver
+	Endpoint string
+	TenantId string
+}
+
+// CreateReplication
+func (v *ReplicationMgr) CreateReplication(body ReplicationBuilder) (*model.ReplicationSpec, error) {
+	var res model.ReplicationSpec
+	url := strings.Join([]string{
+		v.Endpoint,
+		urls.GenerateReplicationURL(urls.Client, v.TenantId)}, "/")
+
+	if err := v.Recv(url, "POST", body, &res); err != nil {
+		return nil, err
+	}
+
+	return &res, nil
+}
+
+// GetReplication
+func (v *ReplicationMgr) GetReplication(replicaId string) (*model.ReplicationSpec, error) {
+	var res model.ReplicationSpec
+	url := strings.Join([]string{
+		v.Endpoint,
+		urls.GenerateReplicationURL(urls.Client, v.TenantId, replicaId)}, "/")
+
+	if err := v.Recv(url, "GET", nil, &res); err != nil {
+		return nil, err
+	}
+
+	return &res, nil
+}
+
+// ListReplications
+func (v *ReplicationMgr) ListReplications() ([]*model.ReplicationSpec, error) {
+	var res []*model.ReplicationSpec
+	url := strings.Join([]string{
+		v.Endpoint,
+		urls.GenerateReplicationURL(urls.Client, v.TenantId, "detail")}, "/")
+
+	if err := v.Recv(url, "GET", nil, &res); err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+// DeleteReplication
+func (v *ReplicationMgr) DeleteReplication(replicaId string, body ReplicationBuilder) error {
+	url := strings.Join([]string{
+		v.Endpoint,
+		urls.GenerateReplicationURL(urls.Client, v.TenantId, replicaId)}, "/")
+	return v.Recv(url, "DELETE", body, nil)
+}
+
+// UpdateReplication
+func (v *ReplicationMgr) UpdateReplication(replicaId string, body ReplicationBuilder) (*model.ReplicationSpec, error) {
+	var res model.ReplicationSpec
+	url := strings.Join([]string{
+		v.Endpoint,
+		urls.GenerateReplicationURL(urls.Client, v.TenantId, replicaId)}, "/")
+
+	if err := v.Recv(url, "PUT", body, &res); err != nil {
+		return nil, err
+	}
+	return &res, nil
+}
+
+// EnableReplication
+func (v *ReplicationMgr) EnableReplication(replicaId string) error {
+	url := strings.Join([]string{
+		v.Endpoint,
+		urls.GenerateReplicationURL(urls.Client, v.TenantId, replicaId, "action")}, "/")
+	body := map[string]interface{}{"enableReplication": map[string]interface{}{}}
```

Comment:
I'm a bit confused about generating an empty body here, how does api server hanle this body? What if we pass it with nil?

Reply:
It's an action url, api controller need using the keyword 'enableReplcation' to distinguish which handler function need to beed invoked.

Reply:
Ok, so is there special mean that specifies the `map[string]interface{}` type?

Reply:
NO, it just guarantees its marshaled string is no difference with josn string which is define in design specs.